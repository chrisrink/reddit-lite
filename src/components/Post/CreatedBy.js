import React from "react";
import PropTypes from "prop-types";
import "./styles/createdby.css";
import moment from "moment";

const humanizeDate = value => moment.unix(value).fromNow();

const propTypes = {
  author: PropTypes.string,
  created: PropTypes.number
};
const CreatedBy = props => {
  return (
    <div className="post-createdby">
      <span>Posted by</span>
      <a
        className="post-author"
        href={`https://reddit.com/user/${props.author}`}
      >
        {`u/${props.author}`}
      </a>
      <span className="post-created">{humanizeDate(props.created)}</span>
    </div>
  );
};

CreatedBy.propTypes = propTypes;

export default CreatedBy;
