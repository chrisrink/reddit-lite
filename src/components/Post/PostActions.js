import React from "react";
import PropTypes from "prop-types";
import "./styles/postactions.css";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import {
  faComment,
  faShare,
  faBookmark
} from "@fortawesome/free-solid-svg-icons";

const propTypes = {
  /**
   * The amount of comments on a post
   */
  commentCount: PropTypes.number,

  /**
   * Triggered when an action is invoked
   * @param {string} actionName
   */
  onAction: PropTypes.func
};

const defaultProps = {
  onAction: () => {},
  commentCount: 0
};
const PostActions = props => {
  const createAction = actionName => () => {
    props.onAction(actionName);

    if (actionName === "comment") {
      window.location.href = props.href;
    }
  };
  const commentText =
    props.commentCount === 0 ? "Comment" : `${props.commentCount} Comments`;

  const handleClick = () => (window.location.href = props.href);

  return (
    <div className="post-actionbar">
      <button className="post-action" onClick={createAction("comment")}>
        <FontAwesomeIcon className="action-icon" icon={faComment} />
        {commentText}
      </button>
      <button className="post-action" onClick={createAction("share")}>
        <FontAwesomeIcon className="action-icon" icon={faShare} />
        Share
      </button>
      <button className="post-action" onClick={createAction("save")}>
        <FontAwesomeIcon className="action-icon" icon={faBookmark} />
        Save
      </button>
    </div>
  );
};

PostActions.propTypes = propTypes;
PostActions.defaultProps = defaultProps;

export default PostActions;
