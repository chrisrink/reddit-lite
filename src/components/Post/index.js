import Post from "./Post";
import "./post.css";

export default Post;
