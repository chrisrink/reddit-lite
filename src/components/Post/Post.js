import React from "react";
import PropTypes from "prop-types";
import PostActions from "./PostActions";
import "./styles/post.css";
import CreatedBy from "./CreatedBy";
import classNames from "classnames";
const propTypes = {
  /**
   * userid of the person who created the post
   */
  author: PropTypes.string,

  /**
   * body text if post is a text type
   */
  selftext: PropTypes.string,

  /**
   * Unix date value since 1970 in ms.
   */
  create: PropTypes.number,

  /**
   * Number of comments on post
   */
  num_comments: PropTypes.number,

  /**
   * Triggered when an action is invoked
   * @param {string} actionName
   */
  onAction: PropTypes.func,

  /**
   * Brief summary of article
   */
  title: PropTypes.string,

  /**
   * If the user has opened the post
   */
  hasVisited: PropTypes.boolean
};

const defaultProps = {
  clicked: false
};
const Post = props => {
  const postClass = classNames({
    post: true,
    [`post-visited`]: props.hasVisited
  });

  const href = `https:\\reddit.com${props.permalink}`;
  return (
    <article className={postClass} style={props.style}>
      <CreatedBy author={props.author} created={props.created_utc} />
      <h3 className="post-title">
        <a href={href}>{props.title}</a>
      </h3>
      {props.selftext && <p className="post-body">{props.selftext}</p>}
      {props.thumbnail && props.thumbnail !== "self" && (
        <div>
          <img
            alt="Post thumbnail"
            className="post-thumbnail"
            src={props.thumbnail}
          />
        </div>
      )}

      <PostActions
        commentCount={props.num_comments}
        onAction={props.onAction}
        href={href}
      />
    </article>
  );
};

Post.propTypes = propTypes;
Post.defaultProps = defaultProps;

export default React.memo(Post);
