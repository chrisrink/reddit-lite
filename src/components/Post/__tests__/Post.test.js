import React from "react";
import { shallow } from "enzyme";
import { Post } from "../Post";

import jsPosts from "../../../api/mock/javascriptList.json";
import epPosts from "../../../api/mock/earthPornList.json";
import CreatedBy from "../CreatedBy";

const jsPost = jsPosts.data.children[3].data;
const epPost = epPosts.data.children[3].data;

describe("Post", () => {
  test("renders a text post", () => {
    expect(shallow(<Post {...jsPost} />).html()).toMatchSnapshot();
  });

  test("renders an image post", () => {
    expect(shallow(<Post {...epPost} />).html()).toMatchSnapshot();
  });

  test("createdBy gets passed the right data", () => {
    const post = shallow(<Post {...jsPost} />);
    expect(post.find(CreatedBy).prop("author")).toEqual("ealmansi");
    expect(post.find(CreatedBy).prop("created")).toEqual(1547493159);
  });

  test("title", () => {
    const post = shallow(<Post {...jsPost} />);
    const title = post.find(".post-title a");
    expect(title.text()).toEqual(
      "TIL: A promise can await itself, thereby creating an infinite (never-resolving) promise."
    );
    expect(title.prop("href")).toEqual(
      "https:\\reddit.com/r/javascript/comments/afz6gp/til_a_promise_can_await_itself_thereby_creating/"
    );
  });
});
