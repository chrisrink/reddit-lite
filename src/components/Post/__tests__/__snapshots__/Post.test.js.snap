// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Post renders a text post 1`] = `

<article class="post">
  <div class="post-createdby">
    <span>
      Posted by
    </span>
    <a class="post-author"
       href="https://reddit.com/user/ealmansi"
    >
      u/ealmansi
    </a>
    <span class="post-created">
      11 days ago
    </span>
  </div>
  <h3 class="post-title">
    <a href="https:\\reddit.com/r/javascript/comments/afz6gp/til_a_promise_can_await_itself_thereby_creating/">
      TIL: A promise can await itself, thereby creating an infinite (never-resolving) promise.
    </a>
  </h3>
  <p class="post-body">
    I&#x27;m currently writing a JavaScript library related to timers and asynchronous code / promises.

I was trying to create a function to stop a recurrently executing process (think setInterval), which returns a promise marking that the process has been successfully stopped. This promise resolves only after the last execution in this process has finished, making it really simple to await until everything has been fully stopped before moving on. This can be useful for writing tests, since you would normally want to avoid recurrent executions from one test lingering on after the test has ended and Mocha/Jest has moved on to the next one.

In code, these are the functions to start off a recurrent execution, and to stop it:

    function loop (fn) {
      var t = { s: false, p: null }
      _loop(fn, t)
      return t
    }
    
    function _loop (fn, t) {
      t.p = !t.s
        ? fn().then(() =&amp;gt; _loop(fn, t))
        : Promise.resolve()
      return t.p
    }
    
    async function stop (t) {
      t.s = true
      await t.p
    }

Then, you can run:

    var t = loop(async () =&amp;gt; {
      await new Promise(r =&amp;gt; setTimeout(r, 1000))
      console.log(&#x27;looping&#x27;)
    })

and you will see the message &quot;looping&quot; show up in the console once every second.

Then, if you call stop, it may go down something like this from a REPL:

    &amp;gt; stop(t).then(() =&amp;gt; console.log(&#x27;fully stopped&#x27;))
    looping
    fully stopped

And then I thought: What happens if I call the function to stop the recurrent process from within one of its executions?

    var t = loop(async () =&amp;gt; {
      await new Promise(r =&amp;gt; setTimeout(r, 1000))
      await stop(t)
    })
    stop(t).then(() =&amp;gt; console.log(&#x27;fully stopped&#x27;))

Surprisingly, the &quot;fully stopped&quot; message is never displayed. Turns out you get a never-ending cycle, since the stop procedure won&#x27;t finish until the last execution finishes, and vice versa.

After finding this out, I started playing around and found many other ways of creating infinite (never-resolving) promises. Here&#x27;s a couple examples that you might find interesting.

You can make a promise await itself:

    var p = Promise.
      resolve().
      then(() =&amp;gt; p).
      then()
    
    var p = (async () =&amp;gt; {
      await Promise.resolve()
      await p
    })()

And you can also make longer cycles, where a promise waits on a promise that waits on itself:

    var p1 = Promise.
      resolve().
      then(() =&amp;gt; p2)
    
    var p2 = p1.then()

It&#x27;s the first time I encountered the concept of promise cycles and found it very interesting, so I thought you guys might also appreciate it.
  </p>
  <div class="post-actionbar">
    <button class="post-action">
      <svg aria-hidden="true"
           data-prefix="fas"
           data-icon="comment"
           class="svg-inline--fa fa-comment fa-w-16 action-icon"
           role="img"
           xmlns="http://www.w3.org/2000/svg"
           viewbox="0 0 512 512"
      >
        <path fill="currentColor"
              d="M256 32C114.6 32 0 125.1 0 240c0 49.6 21.4 95 57 130.7C44.5 421.1 2.7 466 2.2 466.5c-2.2 2.3-2.8 5.7-1.5 8.7S4.8 480 8 480c66.3 0 116-31.8 140.6-51.4 32.7 12.3 69 19.4 107.4 19.4 141.4 0 256-93.1 256-208S397.4 32 256 32z"
        >
        </path>
      </svg>
      16 Comments
    </button>
    <button class="post-action">
      <svg aria-hidden="true"
           data-prefix="fas"
           data-icon="share"
           class="svg-inline--fa fa-share fa-w-16 action-icon"
           role="img"
           xmlns="http://www.w3.org/2000/svg"
           viewbox="0 0 512 512"
      >
        <path fill="currentColor"
              d="M503.691 189.836L327.687 37.851C312.281 24.546 288 35.347 288 56.015v80.053C127.371 137.907 0 170.1 0 322.326c0 61.441 39.581 122.309 83.333 154.132 13.653 9.931 33.111-2.533 28.077-18.631C66.066 312.814 132.917 274.316 288 272.085V360c0 20.7 24.3 31.453 39.687 18.164l176.004-152c11.071-9.562 11.086-26.753 0-36.328z"
        >
        </path>
      </svg>
      Share
    </button>
    <button class="post-action">
      <svg aria-hidden="true"
           data-prefix="fas"
           data-icon="bookmark"
           class="svg-inline--fa fa-bookmark fa-w-12 action-icon"
           role="img"
           xmlns="http://www.w3.org/2000/svg"
           viewbox="0 0 384 512"
      >
        <path fill="currentColor"
              d="M0 512V48C0 21.49 21.49 0 48 0h288c26.51 0 48 21.49 48 48v464L192 400 0 512z"
        >
        </path>
      </svg>
      Save
    </button>
  </div>
</article>

`;

exports[`Post renders an image post 1`] = `

<article class="post">
  <div class="post-createdby">
    <span>
      Posted by
    </span>
    <a class="post-author"
       href="https://reddit.com/user/valentine1"
    >
      u/valentine1
    </a>
    <span class="post-created">
      11 days ago
    </span>
  </div>
  <h3 class="post-title">
    <a href="https:\\reddit.com/r/EarthPorn/comments/afye8a/blue_lake_sits_below_one_of_colorados_most_iconic/">
      Blue Lake sits below one of Colorado&#x27;s most iconic 14ers, and is accessible to everyone as part of Ralph Lauren&#x27;s Ranch. Thanks Ralph. [867x1300] (OC)
    </a>
  </h3>
  <div>
    <img alt="Post thumbnail"
         class="post-thumbnail"
         src="https://b.thumbs.redditmedia.com/3AW-mg1-TsPyQIAkz0WicYC-L5yMPr7aFfLi8Ss3RSI.jpg"
    >
  </div>
  <div class="post-actionbar">
    <button class="post-action">
      <svg aria-hidden="true"
           data-prefix="fas"
           data-icon="comment"
           class="svg-inline--fa fa-comment fa-w-16 action-icon"
           role="img"
           xmlns="http://www.w3.org/2000/svg"
           viewbox="0 0 512 512"
      >
        <path fill="currentColor"
              d="M256 32C114.6 32 0 125.1 0 240c0 49.6 21.4 95 57 130.7C44.5 421.1 2.7 466 2.2 466.5c-2.2 2.3-2.8 5.7-1.5 8.7S4.8 480 8 480c66.3 0 116-31.8 140.6-51.4 32.7 12.3 69 19.4 107.4 19.4 141.4 0 256-93.1 256-208S397.4 32 256 32z"
        >
        </path>
      </svg>
      229 Comments
    </button>
    <button class="post-action">
      <svg aria-hidden="true"
           data-prefix="fas"
           data-icon="share"
           class="svg-inline--fa fa-share fa-w-16 action-icon"
           role="img"
           xmlns="http://www.w3.org/2000/svg"
           viewbox="0 0 512 512"
      >
        <path fill="currentColor"
              d="M503.691 189.836L327.687 37.851C312.281 24.546 288 35.347 288 56.015v80.053C127.371 137.907 0 170.1 0 322.326c0 61.441 39.581 122.309 83.333 154.132 13.653 9.931 33.111-2.533 28.077-18.631C66.066 312.814 132.917 274.316 288 272.085V360c0 20.7 24.3 31.453 39.687 18.164l176.004-152c11.071-9.562 11.086-26.753 0-36.328z"
        >
        </path>
      </svg>
      Share
    </button>
    <button class="post-action">
      <svg aria-hidden="true"
           data-prefix="fas"
           data-icon="bookmark"
           class="svg-inline--fa fa-bookmark fa-w-12 action-icon"
           role="img"
           xmlns="http://www.w3.org/2000/svg"
           viewbox="0 0 384 512"
      >
        <path fill="currentColor"
              d="M0 512V48C0 21.49 21.49 0 48 0h288c26.51 0 48 21.49 48 48v464L192 400 0 512z"
        >
        </path>
      </svg>
      Save
    </button>
  </div>
</article>

`;
